### PR TITLE
Plane: Add support for tailsitter aircraft in quadplanes

### PR DESCRIPTION
**Pull request by tridge**
This adds support for a tailsitter frame type in quadplanes. The controls in tailsitter are taken directly from the fixed wing aircraft.
Perhaps the most significant change is the addition of a AP_AHRS_View class, which allows some parts of the code to see a rotated view of the aircraft attitude. This allows the AC_AttitudeControl code to see a view rotated by 90 degrees in pitch from the view seen by the fixed wing code.
This is a work-in-progress PR. Leonard is looking at ways of improving the throttle control in the AP_MotorsTailsitter class.
A video showing this flying in SITL with RealFlight will be uploaded shortly

### CODE REVIEW

File: ArduPlane/quadplane.h [hunk 1]
```diff
@@ -74,6 +74,15 @@ class QuadPlane
 
     // see if we are flying from vtol point of view
     bool is_flying_vtol(void);
+
+    // return true tailsitter frame configured
```
Comment (OXINARF) [inline ArduPlane/quadplane.h]: Nitpicking: looks to be missing a when

File: ArduPlane/quadplane.h [hunk 2]
```diff
@@ -74,6 +74,15 @@ class QuadPlane
...
+    bool is_tailsitter(void);
+
+    // return true when flying a tailsitter in VOTL
```
Comment (OXINARF) [inline ArduPlane/quadplane.h]: Nitpicking: should it be VTOL?

**Comment by tridge** [comment]
> Wouldn't it be easier if AP_AHRS_View was a subclass of AP_AHRS_NavEKF with a method to change the rotation? I'm pretty sure you thought of that, so can you explain what prevents it?

That would mean duplicating all of the state information inside AP_AHRS_NavEKF, which is quite considerable. I also wanted to do it in a way that didn't add a lot more complexity to AP_AHRS, as it is already pretty complex.

**Comment by OXINARF** [comment]
> That would mean duplicating all of the state information inside AP_AHRS_NavEKF, which is quite considerable.

My idea was to then create that subclass in the vehicles instead of AP_AHRS_NavEKF, having the two would be a no-no obviously.

**Comment by OXINARF** [comment]
Ah forgot one thing: the file `plane-tailsitter.parm` is missing.

**Comment by tridge** [comment]
as a bonus, this PR now also supports dual-motor tailsitters, and as a side effect, supports differential thrust for yaw in fixed wing

**Comment by magicrub** [comment]
Dual motor for tail sitters, or dual motors for all? Full yaw support?
Tridge, you're amazing

**Comment by tridge** [comment]
yes, full dual-motor yaw support for all fixed wing.
It's only a couple of lines of code now we have the SRV_Channel split done

**Comment by tridge** [comment]
I've fixed up all the comments, and removed the todo for heli VFF, as it turns out there is nothing needed for that.
So I think this is ready for master

**Comment by magicrub** [comment]
Closes https://github.com/ArduPilot/ardupilot/issues/1861

**Comment by OXINARF** [comment]
@tridge Have you missed my comments? 

- the commit named "AC_AttitudeControl: added support for AP_AHRS_View" still has the AHRS_View new files there, they should be in the commit named "AP_AHRS: added support for AP_AHRS_View" instead
- the file plane-tailsitter.parm is still missing

**Comment by tridge** [comment]
after successful testing yesterday and today this has been pushed to master

**Comment by TopGun17** [comment]
Helloo!

Nice project tridge! I'm also doing some scool project and i have to built tailsitter/3d hoover plane. Single engine RC plane, same as you have in youtube video.
My question, can i see somewhere the full tailsitter code you showed in video? This would be so helpful to me!

Regards
TopGun

**Comment by magicrub** [comment]
>can i see somewhere the full tailsitter code you showed in video

@TopGun17 You're asking for the source on a thread where he's presenting the source :)
https://github.com/ArduPilot/ardupilot/pull/5708/files

**Pull request closed without merging**